src/rmr_manager.rs

```diff
@@ -0,0 +1,173 @@
+                    let _task = tokio::spawn(mr_timer(timeout, release_rx, timer_tx, token));
+                    debug!("record {:?}", token);
+                    Ok(())
+                },
+                |lmr| {
+                    Err(io::Error::new(
+                        io::ErrorKind::InvalidInput,
+                        format!("the MR {:?} used multiple times", lmr),
+                    ))
+                },
+            )
+    }
+
+    /// Remove mr from map and cancel the corresponding timer task
+    pub(crate) async fn release_mr(&self, token: &MrToken) -> io::Result<LocalMr> {
+        debug!("release_mr {:?}", token);
+        remove_mr_from_map(&self.mr_map, token)
+            .await
+            .map_or_else(Err, |(lmr, release_tx)| {
+                // cancel timer task
+                release_tx.send(()).map_or_else(
+                    |_| {
+                        Err(io::Error::new(
+                            io::ErrorKind::InvalidInput,
+                            format!(
+                                "timer task already droped and the related mr was droped {:?}",
+                                token
+                            ),
+                        ))
+                    },
+                    |_| Ok(lmr),
+                )
+            })
+    }
+}
+
+/// Timer task used to wake up `RemoteMrManager` when the mr timeout
+async fn mr_timer(
+    timeout: Duration,
+    release_rx: oneshot::Receiver<()>,
+    timer_tx: mpsc::Sender<MrToken>,
+    token: MrToken,
+) {
+    let is_timeout = tokio::time::timeout(timeout, release_rx).await.map_or_else(
+        |_| {
+            debug!(
+                "mr_timer is going to wake up rmr_manager to drop timeout mr {:?}",
+                token
+            );
+            true
+        },
+        |_| {
+            debug!(
+                "mr_timer received cancel signal, the corresponding mr is dopped {:?}",
+                token
+            );
+            false
+        },
+    );
+    if is_timeout {
+        timer_tx.send(token).await.map_or_else(
+            |err| {
+                error!("rmr_manager timer receiver droped {:?}", err);
+            },
+            |_| debug!("mr_timer waked up rmr_manager successfully"),
+        );
+    }
+}
+
+/// Remove mr from `RmrMap`
+async fn remove_mr_from_map(
+    map: &RmrMap,
+    token: &MrToken,
+) -> io::Result<(LocalMr, oneshot::Sender<()>)> {
+    map.lock().await.remove(token).map_or_else(
+        || {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("can not find MR {:?}", token),
+            ))
+        },
+        |res| {
+            debug!("removed {:?}", token);
+            Ok(res)
+        },
+    )
+}
+
+/// Main function used to free timeout mrs.
+#[allow(clippy::unreachable)]
+async fn main(mut timer_rx: mpsc::Receiver<MrToken>, map: RmrMap) {
```

Comment:
We'd better not name the function "main", which means the main entry point of an application. Try another name?